src/lib.rs

```diff
@@ -437,3 +438,73 @@ pub mod prelude {
 
     pub use crate::read_only::{ReadOnlyOption, ReadState};
 }
+
+/// Build a logger for tests.
+///
+/// Currently, this is a terminal log. It ensures it is only initialized once to prevent clobbering.
+// This is `pub` so that testing and benching functions can use it.
+// `#[cfg(test)]` doesn't work for benching.
+//
+// TODO: Investigate reporting strategies. Tests are run multithreaded in general, and we'd like to be able to sort through and understand test results.
+#[doc(hidden)]
+pub fn testing_logger() -> &'static Logger {
+    use std::sync::Once;
+    static LOGGER_INITIALIZED: Once = Once::new();
+    static mut LOGGER: Option<Logger> = None;
+
+    unsafe {
+        LOGGER_INITIALIZED.call_once(|| {
+            let decorator = slog_term::TermDecorator::new().build();
+            let drain = slog_term::CompactFormat::new(decorator).build().fuse();
+            let drain = slog_envlogger::new(drain).fuse();
+            let drain = slog_async::Async::new(drain)
+                .chan_size(4096)
+                .overflow_strategy(slog_async::OverflowStrategy::Block)
+                .build()
+                .fuse();
+            LOGGER = Some(slog::Logger::root(drain, o!()));
+        });
+        LOGGER.as_ref().unwrap()
+    }
+}
+
+/// The default logger we fall back to when passed `None` in external facing constructors.
+///
+/// Currently, this is a `log` adaptor behind a `Once` to ensure there is no clobbering.
+#[doc(hidden)]
+fn default_logger() -> &'static Logger {
+    use std::sync::Once;
+    static LOGGER_INITIALIZED: Once = Once::new();
+    static mut LOGGER: Option<Logger> = None;
+
+    unsafe {
+        LOGGER_INITIALIZED.call_once(|| {
+            let drain = slog_stdlog::StdLog.fuse();
+            let drain = slog_envlogger::new(drain).fuse();
+            let drain = slog_async::Async::new(drain)
```

Comment:
Do you think it makes sense to make the default for raft an async logger? If folks don't pass us a logger they're probably using stdlog which would be synchronous. 🤔

Reply:
I think it makes sense to use async. I think if users want the default logger, they will assume they get the best option for common cases (convention over configuration philosophy). If I understand the pros and cons of slog and Raft correctly, then async is usually the better option.

I don't think it is too important to match too closely with other loggers - if they really need that constraint then users should explicitly set the logger.

Reply:
So I thought hard about this, and I'm not sure this is a good idea. If they're not using `slog` and these logging macros are triggering it means they *specifically* **opted in** to syncronous logging with the `log` crate. Using async logs by default could result in out of order logs compared to the relative time of other events in their system (which are acting syncronously).

For example, consider the user is using sync `log` to run their application and are checking when things are happening inside raft to verify their assumptions. Forcing async logging here means they might see raft logs happen *after* the return of their calls, which could create a log of confusion.

I think it's *convention* to opt-in to async logging through explicit configuration. You can see slog itself forces you to opt-in! :)